_source/_posts/2023-02-20-angular-forroot.md

```diff
@@ -0,0 +1,123 @@
+        // make HTTP call to add a cupcake to favorite for this user
+    }
+
+    getFavoriteCupcakes(): Observable<Cupcake[]> {
+        // make HTTP call to get all the favorite cupcakes for this user
+    }
+}
+```
+
+Additionally, you have a component that supports the favorite cupcakes feature in the favorite cupcake module:
+
+```ts
+@NgModule {
+    declarations: [FavoriteCupcakeComponent],
+    exports: [FavoriteCupcakeComponent],
+    providers: [FavoriteCupcakeService]
+}
+export class FavoriteCupcakeModule {}
+```
+
+In your application module, you might import the `FavoriteCupcakeModule`, making the declared `FavoriteCupcakeComponent` and provided `FavoriteCupcakeService` available throughout your application.
+
+But what if you don't import the `FavoriteCupcakeModule` in the `AppModule` or if you want to use the `FavoriteCupcakeComponent` in a lazy-loaded feature module in your app? You'll have to reimport the `FavoriteCupcakeModule` in that lazy-loaded module. Each lazy-loaded module has an injector instance, so you may have multiple instances of the `FavoriteCupcakeService` in your application, which we don't want.
+
+One option is to amend the `@Injectable()` decorator to add the metadata that determines where in the application the service is provided using the `providedIn` property:
+
+```ts
+@Injectable({
+    providedIn: 'root'
+})
+```
+
+This declaration within the decorator is preferable over providing the service within a module's `providers` array because it supports tree-shaking.
+
+Another option is adding the `forRoot()` static method to the `FavoriteCupcakeModule` and providing the `FavoriteCupcakeService`instead of through the `providers` array. The `FavoriteCupcakeService` can and should be a singleton instance for the application, and the `forRoot()` static method sets the module up to do so.
+
+Following in the steps of this best practice, the Okta Angular SDK now has more streamlined configuration options and helps promote good practices of importing the `OktaAuthModule` at the application root. Authentication applies to the entire application and should not depend on where you imported the `OktaAuthModule` in the application.
+
+Let's look at this configuration change.
+
+## Configuring the Okta Angular SDK using OAuth 2.0 and OpenID Connect (OIDC)
+
+Previously, to incorporate Okta in your Angular applications, you had to import the `OktaAuthModule` and pass in your configuration in the `providers` array as a replacement for the `OKTA_CONFIG` injection token like this:
+
+```ts
+const oktaAuth = new OktaAuth({
+    clientId: '{yourOktaClientId}',
+    issuer: 'https://{yourOktaDomain}/oauth2/default'
+});
+
+@NgModule({
+    imports: [ 
+        OktaAuthModule 
+    ],
+    providers: [
+        { provide: OKTA_CONFIG, useValue: {oktaAuth} }
+    ]
+})
+export class AppModule { }
+```
+
+This method will continue to work, and you can continue using the module as you have it now without any changes.
```

Comment:
```suggestion
This method will continue to work, and you can keep on using the module as you have it now without any changes.
```

Reply:
@edunham maybe a fine line, but I'd like to keep it as continue. Trying to encourage the pattern going forward should change, and keep on sounds more positive than continue to me?